File: lib/fortran/benchmark.f90
module benchmark
  implicit none
  private
  public :: run, format_results, benchmark_result_t  ! Make benchmark_result_t public

  type :: benchmark_result_t
    integer :: runs
    real(8) :: mean_ms
    real(8) :: std_dev_ms
    real(8) :: min_ms
    real(8) :: max_ms
    integer(8) :: result
  end type benchmark_result_t

contains

  subroutine run(f, run_ms, result)
    implicit none
    interface
      integer(8) function f()
      end function f
    end interface
    procedure(f), pointer :: func_ptr
    integer, intent(in) :: run_ms
    type(benchmark_result_t), intent(out) :: result
    integer(8) :: start_time, end_time, elapsed_time, total_elapsed_time
    integer(8) :: count_rate
    integer :: count
    real(8) :: elapsed_times(1000000), mean, variance, std_dev, min_time, max_time
    logical :: print_status
    integer(8) :: last_status_t

    ! Check for run_ms being zero
    if (run_ms == 0) then
      result%runs = 0
      result%mean_ms = 0.0
      result%std_dev_ms = 0.0
      result%min_ms = 0.0
      result%max_ms = 0.0
      result%result = 0
      return
    end if

    func_ptr => f
    total_elapsed_time = 0
    count = 0
    min_time = 1.0e12
    max_time = 0.0
    print_status = (run_ms > 1)
    call system_clock(count_rate=count_rate)  ! Get the count rate
    last_status_t = 0

    if (print_status) then
      write(0, '(A)', advance='no') "."
      flush(0)
    end if

    do while (total_elapsed_time < run_ms * 1.0e6)
      if (print_status .and. (total_elapsed_time - last_status_t) > 1.0e9) then
        last_status_t = total_elapsed_time
        write(0, '(A)', advance='no') "."
        flush(0)
      end if
      call system_clock(start_time, count_rate=count_rate)  ! Use nanosecond precision
      result%result = func_ptr()
      call system_clock(end_time, count_rate=count_rate)    ! Use nanosecond precision
      elapsed_time = end_time - start_time
      if (elapsed_time == 0) cycle  ! Skip zero elapsed time measurements
      if (count < size(elapsed_times)) then
        elapsed_times(count + 1) = elapsed_time / 1.0e6
      else
        write(0,*) "Error: Exceeded maximum number of iterations"
        exit
      end if
      total_elapsed_time = total_elapsed_time + elapsed_time
      count = count + 1
      if (elapsed_times(count) < min_time) min_time = elapsed_times(count)
      if (elapsed_times(count) > max_time) max_time = elapsed_times(count)
    end do

    if (print_status) then
      write(0, '(A)') ""
    end if

    mean = sum(elapsed_times(1:count)) / count
    variance = sum((elapsed_times(1:count) - mean)**2) / count
    std_dev = sqrt(variance)

    result%runs = count
    result%mean_ms = mean
    result%std_dev_ms = std_dev
    result%min_ms = min_time
    result%max_ms = max_time
  end subroutine run

  subroutine format_results(benchmark_result, result_str)
      implicit none
      type(benchmark_result_t), intent(in) :: benchmark_result
      character(len=:), allocatable, intent(out) :: result_str
      character(len=256) :: temp_str

      write(temp_str, '(f0.6,",",f0.6,",",f0.6,",",f0.6,",",i0,",",i0)') &
        benchmark_result%mean_ms, benchmark_result%std_dev_ms, benchmark_result%min_ms, benchmark_result%max_ms, benchmark_result%runs, benchmark_result%result

      result_str = trim(adjustl(temp_str))
  end subroutine format_results

end module benchmark

File: lib/julia/benchmark.jl
module Benchmark

export run, format_results

"""
    run(f, run_ms)

Runs the function `f` repeatedly until the total elapsed time exceeds `run_ms` milliseconds.
If `run_ms` is 0, `f` is run once and dummy timing stats are returned.
If `run_ms` is 1, status dots are not printed (this is assumed to be a correctness check).
Returns a NamedTuple with keys:
  - `runs`: number of iterations
  - `result`: the result from the last call to `f`
  - `mean_ms`, `min_ms`, `max_ms`, `std_dev_ms`: timing stats (in milliseconds)
"""
function run(f, run_ms::Real)
    if run_ms == 0
        return (runs = 0, result = f(), mean_ms = 0.0, min_ms = 0.0, max_ms = 0.0, std_dev_ms = 0.0)
    end
    run_ns = run_ms * 1e6  # convert milliseconds to nanoseconds
    init_t = time_ns()
    last_status_t = init_t
    if run_ms > 1
        print(stderr, ".")
        flush(stderr)
    end
    results = Int64[]
    runs = 0
    total_elapsed = 0
    result = nothing
    while total_elapsed < run_ns
        t0 = time_ns()
        result = f()
        t1 = time_ns()
        elapsed = t1 - t0
        push!(results, elapsed)
        total_elapsed += elapsed
        runs += 1
        if run_ms > 1 && (t0 - last_status_t > 1e9)  # print dot if > 1 sec has passed
            print(stderr, ".")
            flush(stderr)
            last_status_t = t0
        end
    end
    if run_ms > 1
        println(stderr)
    end
    mean_ns = total_elapsed / runs
    min_ns = minimum(results)
    max_ns = maximum(results)
    variance = sum((t - mean_ns)^2 for t in results) / runs
    std_dev_ns = sqrt(variance)
    mean_ms = mean_ns / 1e6
    min_ms = min_ns / 1e6
    max_ms = max_ns / 1e6
    std_dev_ms = std_dev_ns / 1e6
    return (runs = runs, result = result, mean_ms = mean_ms, min_ms = min_ms, max_ms = max_ms, std_dev_ms = std_dev_ms)
end

"""
    format_results(stats)

Formats the benchmark results (a NamedTuple produced by `run`) as a comma‑separated string:
`mean_ms,std_dev_ms,min_ms,max_ms,runs,result`
"""
function format_results(stats)
    return string(stats.mean_ms, ",", stats.std_dev_ms, ",", stats.min_ms, ",", stats.max_ms, ",", stats.runs, ",", stats.result)
end

end # module Benchmark
